src/main/java/net/snowflake/ingest/utils/IcebergDataTypeParser.java

```diff
@@ -0,0 +1,160 @@
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.google.common.base.Preconditions;
+import com.google.common.collect.Lists;
+import java.io.IOException;
+import java.util.Iterator;
+import java.util.List;
+import javax.annotation.Nonnull;
+import org.apache.iceberg.types.Type;
+import org.apache.iceberg.types.Types;
+import org.apache.iceberg.util.JsonUtil;
+
+/**
+ * This class is used to Iceberg data type (include primitive types and nested types) serialization
+ * and deserialization.
+ *
+ * <p>This code is modified from
+ * GlobalServices/modules/data-lake/datalake-api/src/main/java/com/snowflake/metadata/iceberg
+ * /IcebergDataTypeParser.java
+ */
+public class IcebergDataTypeParser {
+  private static final String TYPE = "type";
+  private static final String STRUCT = "struct";
+  private static final String LIST = "list";
+  private static final String MAP = "map";
+  private static final String FIELDS = "fields";
+  private static final String ELEMENT = "element";
+  private static final String KEY = "key";
+  private static final String VALUE = "value";
+  private static final String DOC = "doc";
+  private static final String NAME = "name";
+  private static final String ID = "id";
+  private static final String ELEMENT_ID = "element-id";
+  private static final String KEY_ID = "key-id";
+  private static final String VALUE_ID = "value-id";
+  private static final String REQUIRED = "required";
+  private static final String ELEMENT_REQUIRED = "element-required";
+  private static final String VALUE_REQUIRED = "value-required";
+
+  /** Object mapper for this class */
+  private static final ObjectMapper MAPPER = new ObjectMapper();
+
+  /**
+   * Get Iceberg data type information by deserialization.
+   *
+   * @param icebergDataType string representation of Iceberg data type
+   * @return Iceberg data type
+   */
+  public static Type deserializeIcebergType(String icebergDataType) {
+    try {
+      JsonNode json = MAPPER.readTree(icebergDataType);
+      return getTypeFromJson(json);
+    } catch (IOException e) {
+      throw new SFException(ErrorCode.INTERNAL_ERROR, "Failed to deserialize Iceberg data type", e);
+    }
+  }
+
+  /**
+   * Get corresponding Iceberg data type from JsonNode.
+   *
+   * @param jsonNode JsonNode parsed from Iceberg type string.
+   * @return Iceberg data type
+   */
+  public static Type getTypeFromJson(@Nonnull JsonNode jsonNode) {
+    if (jsonNode.isTextual()) {
+      return Types.fromPrimitiveString(jsonNode.asText());
+    } else if (jsonNode.isObject()) {
+      String type = jsonNode.get(TYPE).asText();
+      if (STRUCT.equals(type)) {
+        return structFromJson(jsonNode);
+      } else if (LIST.equals(type)) {
+        return listFromJson(jsonNode);
+      } else if (MAP.equals(type)) {
+        return mapFromJson(jsonNode);
+      }
+    }
+
+    throw new SFException(ErrorCode.INTERNAL_ERROR, "Cannot parse Iceberg type from: " + jsonNode);
```

Comment:
1. instead of reserializing jsonNode I'd pass in the original string and write that out to logs / to the error message.

2. (P2) It might be a good idea to only throw SFExceptions from "higher" layers and not from utility parsing classes like this one, IllegalArgumentException probably is good enough here; lets chat with Toby on whether there's already a defined scheme to which all levels can/should throw SFException.

Reply:
Changed all exception to `IllegalArgurmentException` in this class.